File: swift/ConcertoTests/HostedTestAppTerminationObserver.m
#import <AppKit/AppKit.h>
#import <XCTest/XCTest.h>

@interface HostedTestAppTerminationObserver : NSObject <XCTestObservation>
@end

@implementation HostedTestAppTerminationObserver

- (void)testBundleDidFinish:(NSBundle *)testBundle
{
    dispatch_async(dispatch_get_main_queue(), ^{
        if (NSApp == nil) {
            return;
        }

        fprintf(stderr, "ConcertoTests: terminating hosted app after test bundle finished\n");
        fflush(stderr);
        [NSApp terminate:nil];
    });
}

@end

static HostedTestAppTerminationObserver *hostedTestAppTerminationObserver = nil;

__attribute__((constructor))
static void InstallHostedTestAppTerminationObserver(void)
{
    hostedTestAppTerminationObserver = [HostedTestAppTerminationObserver new];
    [[XCTestObservationCenter sharedTestObservationCenter] addTestObserver:hostedTestAppTerminationObserver];
}

File: swift/ConcertoTests/HostedTestAppTerminationObserver.swift
import XCTest

#if os(macOS)
import AppKit
import Foundation

private final class HostedTestAppTerminationObserver: NSObject, XCTestObservation, @unchecked Sendable {
    func testBundleDidFinish(_ testBundle: Bundle) {
        Task { @MainActor in
            guard NSApp != nil else { return }
            FileHandle.standardError.write(Data("ConcertoTests: terminating hosted app after test bundle finished\n".utf8))
            NSApp.terminate(nil)
        }
    }
}

private let hostedTestAppTerminationObserver: HostedTestAppTerminationObserver = {
    let observer = HostedTestAppTerminationObserver()
    XCTestObservationCenter.shared.addTestObserver(observer)
    return observer
}()
#endif
